crates/core/src/host/wasmer/wasm_instance_env.rs

```diff
@@ -35,9 +47,79 @@ fn mem_err(err: MemoryAccessError) -> RuntimeError {
     }
 }
 
+/// Evaluate `body` with WASM execution timing paused.
+///
+/// `caller` should be an ident bound mutably to a `FunctionEnvMut<'_, WasmInstanceEnv>`.
+///
+/// `body` may be arbitrary code, including uses of `return` or `?`,
+/// except that panicing inside the `body` and then catching the panic outside the body
+/// will cause the reported time to be incorrect.
+///
+/// Returns from `body` will be intercepted, so that timing can be resumed after it exits.
+/// This macro will always return the result of `body`, even if `body` does an apparently non-local exit,
+/// like `return` or `?`.
+///
+/// This might be better done with an RAII gaurd,
+/// except that version would mutably borrow the `caller` for the duration of `body`,
+/// preventing the `body` from using the `caller`.
+/// This version mutably borrows `caller` once before `body`, and again after,
+/// but does not hold the borrow during `body`.
+macro_rules! with_timing_paused {
+    ($caller:ident -> { $($body:tt)* }) => {{
+        $caller.data_mut().timing_after_exit_wasm();
+
+        // Improvised try block: errors bubbled with `?` are captured in `res`,
+        // and the `timing_before_enter_wasm` runs regardless.
+        //
+        // Clippy will warn about this if `body` doesn't include a `?` or `return`,
```

Comment:
I'm no expert at the interactions of clippy and macros but this seems fine to me.